examples/static_set/mapping_table_example.cu

```diff
@@ -0,0 +1,154 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuco/static_set.cuh>
+
+#include <cuda/std/array>
+#include <cuda/std/tuple>
+#include <thrust/device_vector.h>
+
+#include <iostream>
+
+/**
+ * @file mapping_table_example.cu
+ *
+ * @brief Demonstrates how to use hash set as a lookup table of the original data
+ *
+ * `cuco` hash tables such as `cuco::static_set` or `cuco::static_map` currently support only 4/8
+ * byte keys. This limitation arises because `cuco` hash tables rely on atomic Compare-And-Swap
+ * (CAS) operations for key insertions (or queries), and the hardware natively supports only 4-byte
+ * and 8-byte CAS. To enable support for larger keys, one approach is to implement atomic lock
+ * tables at the software level. However, this approach would lead to a notable performance decrease
+ * due to the high runtime cost of atomic lock tables.
+ *
+ * Additionally, `cuco` hash tables use open addressing as the hash collision resolution method.
+ * This approach requires users to provide a sentinel that indicates unused slots in the data
+ * structure. The sentinel value is a reserved value that must be never present in the problem. Note
+ * that inserting or querying a sentinel value is undefined behavior. This can be problematic,
+ * especially when the input data type is complex and determining a valid sentinel value is not
+ * straightforward.
+ *
+ * This sample code demonstrates a solution to address these issues by using hash set as an
+ * indirection mapping table to the original data:
+ *  - The keys inserted in the hash table are indices of the original data array.
+ *  - Using `-1` as a sentinel value is safe because accessing `data[-1]` is invalid.
+ *  - Custom hashers and key equality comparators are required to hash and compare original keys
+ *    based on indices.
+ *
+ * @note This example is for demonstration purposes only. It is not intended to show the most
+ * performant way to do the example algorithm.
+ */
+
+/**
+ * @brief User-defined key equal to compare two keys
+ *
+ * @tparam T Original key type
+ */
+template <typename T>
+struct my_equal {
+  my_equal(T const* data) : _data{data} {}
+  /**
+   * @brief Checks if two keys are identical based on their indices in the
+   * original data array
+   *
+   * @param lhs The left hand side index
+   * @param rhs The right hand side index
+   * @return 'true' if two tuples are identical
+   */
+  __device__ constexpr bool operator()(int32_t lhs, int32_t rhs) const
+  {
+    // Check all 4 elements of a tuple to determine if two tuples are identical
+    return cuda::std::get<0>(_data[lhs]) == cuda::std::get<0>(_data[rhs]) and
+           cuda::std::get<1>(_data[lhs]) == cuda::std::get<1>(_data[rhs]) and
+           cuda::std::get<2>(_data[lhs]) == cuda::std::get<2>(_data[rhs]) and
+           cuda::std::get<3>(_data[lhs]) == cuda::std::get<3>(_data[rhs]);
+  }
+  T const* _data;
```

Comment:
Should we consider using `cuda::std::span` here?

Reply:
Using pointer for demonstration purposes since not everyone is familiar with the span concept